File: firmware/wt32_sc01_plus/ui/machine_position.py
import lvgl as lv
from lv_style import *
import gc

class MachinePositionWCS(lv.obj):
    FONT_WIDTH = 11
    FONT_HEIGHT = 22
    DEFAULT_COORD_SYSTEMS = ['Mach', 'WCS', 'Move']

    def __init__(self, parent, coords, interface, digits = 5,
                 coord_systems=DEFAULT_COORD_SYSTEMS, coord_sys_width = 40):
        super().__init__(parent)

        self.interface = interface
        self.parent = parent
        self.coords = coords
        self.digits = digits
        self.fmt_str = "%%%d.2f" % (digits - 2)
        self.coord_systems = coord_systems

        container = self
        style_container_blank(self)
        container.set_style_pad_column(2, lv.STATE.DEFAULT)
        container.set_style_pad_row(2, lv.STATE.DEFAULT)
        container.set_layout(lv.LAYOUT.GRID)
        non_scrollable(container)
        no_margin_pad_border(container)

        font = self.interface.font_lcd_24

        gc.collect()

        lblc_width = (digits + 1) * self.FONT_WIDTH
        lencs = len(coord_systems)
        lenc = len(coords)
        col_dsc = [coord_sys_width] + [lblc_width] * lencs + [lv.GRID_TEMPLATE_LAST]
        row_dsc = [self.FONT_HEIGHT + 2] * (lenc + 1) + [lv.GRID_TEMPLATE_LAST]

        container.set_style_grid_column_dsc_array(col_dsc, 0)
        container.set_style_grid_row_dsc_array(row_dsc, 0)
        container.set_size(coord_sys_width + lencs * 2 + lblc_width * lencs + 2,
                           lv.SIZE_CONTENT)

        self.axis_label_ids = {}
        self.axis_labels = []
        # Create Axis labels.
        for i, coord in enumerate(coords):
            lbl = lv.label(container)
            lbl.set_text(lv.SYMBOL.HOME + ' ' + coord)
            style(lbl, { 'bg_opa': 100, 'margin': 1, 'padding': 0 })
            lbl.set_grid_cell(lv.GRID_ALIGN.STRETCH, 0, 1,
                              lv.GRID_ALIGN.STRETCH, i + 1, 1)
            lbl.set_style_text_align(lv.TEXT_ALIGN.CENTER, lv.STATE.DEFAULT)
            lbl.center()
            lbl.add_event_cb(self._label_home_clicked, lv.EVENT.CLICKED, None)
            lbl.add_flag(lv.obj.FLAG.CLICKABLE)
            self.axis_label_ids[lbl.get_text()] = coord
            self.axis_labels.append(lbl)
            colr = color('BLUE') if interface.machine.axes_homed[i] else color('RED')
            lbl.set_style_bg_color(colr, lv.STATE.DEFAULT)

        # Coordinate systems.
        self.coord_sys_labels = []
        for i, cs in enumerate(coord_systems):
            print(i, cs)
            lbl = lv.label(container)
            lbl.set_text(cs)
            style(lbl, { 'bg_color': color('BLUE') , 'bg_opa': 100,
                         'margin': 1, 'padding': 0 })
            lbl.set_grid_cell(lv.GRID_ALIGN.STRETCH, i + 1, 1,
                              lv.GRID_ALIGN.STRETCH, 0, 1)
            lbl.set_style_text_align(lv.TEXT_ALIGN.CENTER, lv.STATE.DEFAULT)
            lbl.center()
            self.coord_sys_labels.append(lbl)

        self.coord_vals = {}
        self.coord_val_labels = {}
        self.coord_val_labels_by_id = []

        for i, cs in enumerate(coord_systems):
            c_labels = []
            self.coord_val_labels[cs] = c_labels
            self.coord_val_labels_by_id.append(cs)
            self.coord_vals[cs] = [None] * len(self.coords)

            for j, coord in enumerate(coords):
                lblc = lv.label(container)
                lblc.set_text('?' * (digits - 2) + '.??')
                style(lblc, { 'bg_color': color('LIME') , 'bg_opa': 60,
                             'margin': 0, 'padding': [2, 2] })

                if font is not None:
                    lblc.set_style_text_font(font, lv.STATE.DEFAULT)

                lblc.set_width(lblc_width)
                lblc.set_grid_cell(lv.GRID_ALIGN.STRETCH, i + 1, 1,
                                  lv.GRID_ALIGN.STRETCH, j + 1, 1)
                lblc.center()
                lblc.set_style_text_align(lv.TEXT_ALIGN.RIGHT, lv.STATE.DEFAULT)
                c_labels.append(lblc)

        interface.machine.add_pos_changed_cb(self._pos_updated)
        interface.machine.add_home_changed_cb(self._home_updated)
        interface.machine.add_wcs_changed_cb(self._wcs_updated)

        self._wcs_updated(interface.machine)

    def coords_undefined(self):
        for cs in self.coord_systems:
            self.coord_vals[cs] = [None] * len(self.coords)
            for lblc in self.coord_val_labels[cs]:
                lblc.set_text('?' * (self.digits - 2) + '.??')

    def set_coord(self, c, v, coord_system=None):
        if coord_system is None:
            coord_system = self.coord_systems[0]

        if isinstance(c, str):
            c = self.coords.index(c)

        if not self.interface.machine.axes_homed[c]:
            v = None

        if v != self.coord_vals[coord_system][c]:
            self.coord_vals[coord_system][c] = v
            if v is not None:
                self.coord_val_labels[coord_system][c].set_text(self.fmt_str % v)
            else:
                self.coord_val_labels[coord_system][c].set_text('?' * (self.digits - 2) + '.??')

    def _label_home_clicked(self, e):
        label = lv.label.__cast__(e.get_target())
        ax = self.axis_label_ids[label.get_text()]
        self.interface.machine.home(ax)

    def _pos_updated(self, mach):
        for i, coord in enumerate(mach.position):
            self.set_coord(i, coord, self.coord_systems[0])
        for i, coord in enumerate(mach.wcs_position):
            self.set_coord(i, coord, self.coord_systems[1])

    def _home_updated(self, mach):
        for i, homed in enumerate(mach.axes_homed):
            colr = color('BLUE') if homed else color('RED')
            self.axis_labels[i].set_style_bg_color(colr, lv.STATE.DEFAULT)
        self._pos_updated(mach)

    def _wcs_updated(self, mach):
        t = 'G%d' % (mach.wcs + 53)
        self.coord_sys_labels[1].set_text(t)

class MachinePosition:
    FONT_WIDTH = 7

    def __init__(self, parent, coords, interface, digits = 5):
        self.interface = interface
        self.parent = parent
        self.coords = coords
        self.digits = digits
        self.fmt_str = "%%%d.2f" % (digits - 2)

        container = create_container(parent)
        self.container = container
        flex_row(container)
        container.set_style_pad_column(0, lv.STATE.DEFAULT)
        non_scrollable(container)

        self.coord_vals = [None, None, None]

        self.coord_labels = []
        font = self.interface.font_lcd_18 or self.interface.font_lcd
        for coord in coords:
            lbl = lv.label(container)
            lbl.set_text(coord + ' ')
            style(lbl, { 'bg_color': color('BLUE') , 'bg_opa': 100,
                         'margin': 1, 'padding': 0 })

            lblc = lv.label(container)
            lblc.set_text('?' * (digits - 2) + '.??')
            style(lblc, { 'bg_color': color('LIME') , 'bg_opa': 100,
                         'margin': 0, 'padding': [2, 0] })


            if font is not None:
                lblc.set_style_text_font(font, lv.STATE.DEFAULT)
            lblc.set_width((digits + 1) * self.FONT_WIDTH)

            self.coord_labels.append(lblc)

    def coords_undefined(self):
        self.coord_vals = [None, None, None]
        for lblc in self.coord_labels:
            lblc.set_text('?' * (self.digits - 2) + '.??')

    def set_coord(self, c, v):
        if isinstance(c, str):
            c = self.coords.index(c)
        if v != self.coord_vals[c]:
            self.coord_labels[c].set_text(self.fmt_str % v)
            self.coord_vals[c] = v

File: firmware/wt32_sc01_plus_pio/src/ui/machine_position.cpp
#include "machine_position.hpp"
#include <stdio.h> // For sprintf
#include <lvgl.h>
#include "ui/interface.hpp"
#include "machine/machine_interface.hpp"

#include "ui/lv_style.hpp"

const std::vector<std::string> MachinePositionWCS::DEFAULT_COORD_SYSTEMS = {"Mach", "WCS", "Move"};

// Implement the color function here, or in a common utility file
int color(const char* name) {
    if (strcmp(name, "BLUE") == 0) return st_color(lv_color_hex(0x0000FF));
    if (strcmp(name, "RED") == 0) return st_color(lv_color_hex(0xFF0000));
    if (strcmp(name, "LIME") == 0) return st_color(lv_color_hex(0x00FF00));
    // Add more colors as needed
    return st_color(lv_color_hex(0x808080)); // Default to gray
}

MachinePositionWCS::MachinePositionWCS(lv_obj_t* parent, const std::vector<std::string>& coords,
                                       Interface* interface, int digits,
                                       const std::vector<std::string>& coord_systems,
                                       int coord_sys_width) {
    this->self = lv_obj_create(parent);
    this->interface = interface;
    this->parent = parent;
    this->coords = coords;
    this->digits = digits;
    char fmt[20];
    sprintf(fmt, "%%%d.2f", digits - 2);
    this->fmt_str = fmt;
    this->coord_systems = coord_systems;

    lv_obj_t* container = self;
    style_container_blank(self);
    lv_obj_set_style_pad_column(container, 2, LV_STATE_DEFAULT);
    lv_obj_set_style_pad_row(container, 2, LV_STATE_DEFAULT);
    lv_obj_set_layout(container, LV_LAYOUT_GRID);
    non_scrollable(container);
    no_margin_pad_border(container);

    lv_font_t* font = this->interface->font_lcd_24; // Assuming this is how you access the font

    int lblc_width = (digits + 1) * FONT_WIDTH;
    int lencs = coord_systems.size();
    int lenc = coords.size();

    static lv_coord_t col_dsc[10]; // Adjust size as needed for your grid
    col_dsc[0] = coord_sys_width;

    for (int i = 1; i <= lencs; ++i) {
        col_dsc[i] = lblc_width;
    }
    col_dsc[lencs + 1] = LV_GRID_TEMPLATE_LAST;

    static lv_coord_t row_dsc[10]; // Adjust size as needed for your grid
    for (int i = 0; i <= lenc; ++i) {
        row_dsc[i] = FONT_HEIGHT + 2;
    }
    row_dsc[lenc + 1] = LV_GRID_TEMPLATE_LAST;

    lv_obj_set_style_grid_column_dsc_array(container, col_dsc, 0);
    lv_obj_set_style_grid_row_dsc_array(container, row_dsc, 0);
    lv_obj_set_size(container, coord_sys_width + lencs * 2 + lblc_width * lencs + 2, LV_SIZE_CONTENT);


    // Create Axis labels.
    for (size_t i = 0; i < coords.size(); ++i) {
        lv_obj_t* lbl = lv_label_create(container);
        std::string text = LV_SYMBOL_HOME " " + coords[i];
        lv_label_set_text(lbl, text.c_str());
        style(lbl, {{"bg_opa", 100}, {"margin", 1}, {"padding", 0}});
        lv_obj_set_grid_cell(lbl, LV_GRID_ALIGN_STRETCH, 0, 1, LV_GRID_ALIGN_STRETCH, i + 1, 1);
        lv_obj_set_style_text_align(lbl, LV_TEXT_ALIGN_CENTER, LV_STATE_DEFAULT);
        lv_obj_center(lbl);
        lv_obj_add_event_cb(lbl, [](lv_event_t* e) {
            MachinePositionWCS *self = static_cast<MachinePositionWCS*>(lv_event_get_user_data(e));
            self->_label_home_clicked(e);
        }, LV_EVENT_CLICKED, this);

        lv_obj_add_flag(lbl, LV_OBJ_FLAG_CLICKABLE);
        axis_label_ids[text] = coords[i];
        axis_labels.push_back(lbl);
        lv_color_t colr = t_color(interface->machine->axes_homed[i] ? color("BLUE") : color("RED"));
        lv_obj_set_style_bg_color(lbl, colr, LV_STATE_DEFAULT);
    }


    // Coordinate systems.
    for (size_t i = 0; i < coord_systems.size(); ++i) {
        lv_obj_t* lbl = lv_label_create(container);
        lv_label_set_text(lbl, coord_systems[i].c_str());
        style(lbl, {{"bg_color", color("BLUE")}, {"bg_opa", 100}, {"margin", 1}, {"padding", 0}});
        lv_obj_set_grid_cell(lbl, LV_GRID_ALIGN_STRETCH, i + 1, 1, LV_GRID_ALIGN_STRETCH, 0, 1);
        lv_obj_set_style_text_align(lbl, LV_TEXT_ALIGN_CENTER, LV_STATE_DEFAULT);
        lv_obj_center(lbl);
        coord_sys_labels.push_back(lbl);

        if (i == 1) {
            lv_obj_add_event_cb(lbl, [](lv_event_t* e) {
                MachinePositionWCS* self = (MachinePositionWCS*) lv_event_get_user_data(e);
                self->interface->machine->nextWcs();
            }, LV_EVENT_CLICKED, this);
            lv_obj_add_flag(lbl, LV_OBJ_FLAG_CLICKABLE);
        }
    }

    // Coordinate Values
    for (size_t i = 0; i < coord_systems.size(); ++i) {
        std::vector<lv_obj_t*> c_labels;
        coord_val_labels[coord_systems[i]] = c_labels;
        coord_val_labels_by_id.push_back(coord_systems[i]);
        coord_vals[coord_systems[i]] = std::vector<float>(coords.size(), NAN); // Initialize with NAN


        for (size_t j = 0; j < coords.size(); ++j) {
            lv_obj_t* lblc = lv_label_create(container);
            std::string q_marks(digits - 2, '?');
            lv_label_set_text(lblc, (q_marks + ".??").c_str());

            style(lblc, {{"bg_color", color("LIME")}, {"bg_opa", 60}, {"margin", 0}, {"padding", 2}});


            if (font != nullptr) {
                lv_obj_set_style_text_font(lblc, font, LV_STATE_DEFAULT);
            }
            lv_obj_set_width(lblc, lblc_width);
            lv_obj_set_grid_cell(lblc, LV_GRID_ALIGN_STRETCH, i + 1, 1, LV_GRID_ALIGN_STRETCH, j + 1, 1);
            lv_obj_center(lblc);
            lv_obj_set_style_text_align(lblc, LV_TEXT_ALIGN_RIGHT, LV_STATE_DEFAULT);
            coord_val_labels[coord_systems[i]].push_back(lblc);
        }
    }

    interface->machine->addPosChangedCb([this](MachineInterface* mach) { _pos_updated(mach); });
    interface->machine->addHomeChangedCb([this](MachineInterface* mach) { _home_updated(mach); });
    interface->machine->addWcsChangedCb([this](MachineInterface* mach) { _wcs_updated(mach); });

    _wcs_updated(interface->machine);
}

void MachinePositionWCS::coordsUndefined() {
    for (const auto& cs : coord_systems) {
        coord_vals[cs] = std::vector<float>(coords.size(), NAN);
        for (lv_obj_t* lblc : coord_val_labels[cs]) {
            std::string q_marks(digits - 2, '?');
            lv_label_set_text(lblc, (q_marks + ".??").c_str());
        }
    }
}

void MachinePositionWCS::setCoord(int ax, float v, const std::string& coord_system) {
    std::string cs = coord_system.empty() ? coord_systems[0] : coord_system;

    if (ax < 0 || ax >= coords.size()) {
        Serial.println("Axis index out of bounds!");
        return;
    }

    if (!interface->machine->axes_homed[ax]) {
        v = NAN;
    }

    if (isnan(v) && isnan(coord_vals[cs][ax])) return;  // both are NAN, no update needed
    if (!isnan(v) && !isnan(coord_vals[cs][ax]) && abs(v - coord_vals[cs][ax]) < 0.0001) return; //values are close enough

    coord_vals[cs][ax] = v;

    char buf[32];
    if (!isnan(v)) {
        snprintf(buf, sizeof(buf), fmt_str.c_str(), v);
        lv_label_set_text(coord_val_labels[cs][ax], buf);
    } else {
        std::string q_marks(digits - 2, '?');
        lv_label_set_text(coord_val_labels[cs][ax], (q_marks + ".??").c_str());
    }
}

void MachinePositionWCS::setCoord(const std::string& ax, float v, const std::string& coord_system) {
    auto it = std::find(coords.begin(), coords.end(), ax);
    if (it != coords.end()) {
        setCoord(std::distance(coords.begin(), it), v, coord_system);
    } else {
        Serial.println("Invalid axis name!");
    }
}


void MachinePositionWCS::_label_home_clicked(lv_event_t* e) {
    lv_obj_t* label = (lv_obj_t *) lv_event_get_target(e);
    std::string ax_text = lv_label_get_text(label);
    std::string ax = axis_label_ids[ax_text];
    interface->machine->home(ax);
}

void MachinePositionWCS::_pos_updated(MachineInterface* mach) {
    // TODO: Machine, WCS and remaining coord systems currently hard-coded.
    for (size_t i = 0; i < mach->position.size(); ++i) {
        setCoord(i, mach->position[i], coord_systems[0]);
    }
    if (coord_systems.size() > 1) {
        for (size_t i = 0; i < mach->wcs_position.size(); ++i) {
            setCoord(i, mach->wcs_position[i], coord_systems[1]);
            float pos = mach->position[i];
            float coord = mach->wcs_position[i];
            if (coord_systems.size() > 2 && !isnan(pos) && !isnan(coord)) {
                float diff = coord - pos;
                setCoord(i, diff, coord_systems[2]);
            }
        }
    }
}

void MachinePositionWCS::_home_updated(MachineInterface* mach) {
    for (size_t i = 0; i < mach->axes_homed.size(); ++i) {
        lv_color_t colr = t_color(mach->axes_homed[i] ? color("BLUE") : color("RED"));
        lv_obj_set_style_bg_color(axis_labels[i], colr, LV_STATE_DEFAULT);
    }
    _pos_updated(mach);
}

void MachinePositionWCS::_wcs_updated(MachineInterface* mach) {
    std::string t = mach->getWcsStr();
    lv_label_set_text(coord_sys_labels[1], t.c_str());
}


MachinePosition::MachinePosition(lv_obj_t* parent, const std::vector<std::string>& coords,
                       Interface* interface, int digits) {
        this->self = lv_obj_create(parent);
        this->interface = interface;
        this->parent = parent;
        this->coords = coords;
        this->digits = digits;
        char fmt[20];
        sprintf(fmt, "%%%d.2f", digits - 2);
        this->fmt_str = fmt;

        this->container = create_container(parent);
        flex_row(container, 0, 0, false);
        lv_obj_set_style_pad_column(container, 0, LV_STATE_DEFAULT);
        non_scrollable(container);

        coord_vals = std::vector<float>(coords.size(), NAN);

        lv_font_t* font = this->interface->font_lcd_18 ? this->interface->font_lcd_18 : this->interface->font_lcd;
        for (const auto& coord : coords) {
            lv_obj_t* lbl = lv_label_create(container);
            std::string text = coord + " ";
            lv_label_set_text(lbl, text.c_str());
            style(lbl, {{"bg_color", color("BLUE")}, {"bg_opa", 100}, {"margin", 1}, {"padding", 0}});

            lv_obj_t* lblc = lv_label_create(container);
            std::string q_marks(digits - 2, '?');
            lv_label_set_text(lblc, (q_marks + ".??").c_str());
            style(lblc, {{"bg_color", color("LIME")}, {"bg_opa", 100}, {"margin", 0}, {"padding", 2}});

            if (font != nullptr) {
                lv_obj_set_style_text_font(lblc, font, LV_STATE_DEFAULT);
            }
            lv_obj_set_width(lblc, (digits + 1) * FONT_WIDTH);

            coord_labels.push_back(lblc);
        }
}


void MachinePosition::coordsUndefined(){
    coord_vals = std::vector<float>(coords.size(), NAN);
    for (lv_obj_t* lblc : coord_labels) {
            std::string q_marks(digits - 2, '?');
            lv_label_set_text(lblc, (q_marks + ".??").c_str());
    }
}


void MachinePosition::setCoord_(int c, float v) {
    if (c < 0 || c >= coords.size()) {
        Serial.println("Axis index out of bounds!");
        return;
    }

    if (isnan(v) && isnan(coord_vals[c])) return;  // both are NAN, no update needed
    if (!isnan(v) && !isnan(coord_vals[c]) && abs(v - coord_vals[c]) < 0.0001) return; //values are close enough

    coord_vals[c] = v;

    char buf[32];
    if (!isnan(v)) {
        snprintf(buf, sizeof(buf), fmt_str.c_str(), v);
        lv_label_set_text(coord_labels[c], buf);
    } else {
        std::string q_marks(digits - 2, '?');
        lv_label_set_text(coord_labels[c], (q_marks + ".??").c_str());
    }
}

void MachinePosition::setCoord_(const std::string& c, float v) {
        auto it = std::find(coords.begin(), coords.end(), c);
    if (it != coords.end()) {
        setCoord_(std::distance(coords.begin(), it), v);
    } else {
        Serial.println("Invalid axis name!");
    }
}